Start User: Add Continue as user to the sign in flow

enejb opened the pull request:
Instead of showing a notice that is not very helpful when the user clicks the browser back button on the domain screen.

Let's show the user a Continue as User component instead.

#### Changes proposed in this Pull Request

When the "user has completed the registration step" has been detected. We redirect the user to the registration form (same url) but with the ?user_completed=1 query parameter in the URL that helps us know if a user really want to be on this page. Usually, we skip the user registration step if they are already logged in. 


#### Testing instructions
See 

https://user-images.githubusercontent.com/115071/142338815-70a426a7-2ce9-4646-8e10-95ac44d16af6.mp4

* Visit 
/start 
Create a new user. 
Once you are on the domain Screen. Use the browser back button. 

Notice that you are logged in as a newly created user. 
* Click continue - you should end up on the domain screen again
* Click the browser back button. 
* Click Login in as someone else 
- notice that you end up on the /start/user screen. 
- You are now logged out. 

Other things to test
visit /log-in and /log-in/new notice that it looks as expected.

Related to https://github.com/Automattic/wp-calypso/issues/57287

enejb committed:
Start User: Add Continue as user to the sign in flow

Instead of showing a notice this lets is not very helpful
Let's show the user a Continue as User component.

annchichi commented:
Hey @enejb in the video 0:27, just want to double-check here:
-  it will take the user back to this screen then auto direct to the "Let's Get Started - Continue" screen. Or we can just take the user directly to the "Let's Get Started - Continue" screen without showing the form in 0:27? 
- Let's get started copy should be centered. 
- Let's get started copy and Avatar spacing should be about 48px. Current one is about 30px. 

If I spot anything, will let you know. Thank you! 😺

enejb commented:
Yes I am hoping to remove the screen that shows up on 0:27. 

I still need to play around a bit to try to figure that part out. @tyxla do you have any suggestions on how to fix that? 

> Let's get started copy should be centred.
I mostly keep the heading not centred mostly since on mobile most of the other screens (domains). Plans is centered agina. 

I think we can probably fix that across all the screens on mobile. Would that make sense to you? @annchichi 

> Let's get started copy and Avatar spacing should be about 48px. Current one is about 30px.

Good catch!

tyxla commented on client/signup/steps/user/index.jsx:
```diff
@@ -447,6 +451,17 @@ export class UserStep extends Component {
 			isSocialSignupEnabled = true;
 		}
 
+		if ( this.props.step && 'completed' === this.props.step.status ) {
```
Rendering doesn't seem like the right time to perform such a side effect. At the time the redirect happens, the redirect will have finished, which is likely why you're seeing the form shortly before the redirect.

Have you tried moving the redirection logic to `componentDidUpdate()`? Furthermore, it might need to be moved above to the controller level, since it happening in the step component could already be "too late".

enejb committed:
Bump the spacing to 48px between the heading and text

enejb committed:
Center the Heading on mobile on sign up

enejb committed:
Only redirect if the user has used History Navigation

This is done so that we don't accidentally send the user to the continue page when they are creating a new user account

enejb committed:
Fix firefox issue with the back button

On second try in Firefox the back button didn't work any more.
this fixes it by making sure that we don't remove the sign in step if there the page as loaded via history.

enejb commented:
@tyxla I managed to get things working in FF now. 🥳 
Here is a short video. 

https://user-images.githubusercontent.com/115071/143457593-da6aefe7-76ca-4ab5-9060-7b6d06ba7a60.mp4

Can you take a look again?

 I also came across an error in the `RegisterDomainStep` 
```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
RegisterDomainStep@app:///./components/domains/register-domain-step/index.jsx:160:5
_class@app:///../packages/i18n-calypso/src/localize.js:36:9
ShoppingCartWrapper@app:///../packages/shopping-cart/src/with-shopping-cart.tsx:15:41
CartKeyWrapper@app:///./my-sites/checkout/with-cart-key.tsx:13:78
ConnectFunction@app:///../node_modules/react-redux/es/components/connectAdvanced.js:233:68
ShoppingCartProvider@app:///../packages/shopping-cart/src/shopping-cart-provider.tsx:18:30
CalypsoShoppingCartProvider@app:///./my-sites/checkout/calypso-shopping-cart-provider.tsx:20:37
div
```

That I think we should address in a different PR.

annchichi commented:
Woohoo! 0:44 is working now. Thanks @enejb 🥳 

Regarding the copy "Log in with another account", it was not clear if users can also sign out based on design feedback. So I think we can be even more clearer like this "Sign out or log in with another account" What do you think? 😄 

Thanks for this!

enejb commented:
It is definitely something we can update. I think a more refined wording makes sense here.

tyxla commented on client/blocks/login/continue-as-user.scss:
```diff
@@ -1,34 +1,68 @@
 	color: var( --color-text-subtle );
 	font-size: $font-body-small;
 	text-align: center;
+	height: calc( 70vh - 80px );
```
This seems a bit made-up, can we add a comment as to where it came from?

enejb replied:
You are right. I am not really sure exactly why those numbers have been chosen but they make things look okay across different screen sizes that I have tested.

tyxla commented on client/blocks/login/continue-as-user.scss:
```diff
@@ -1,34 +1,68 @@
+	.continue-as-user__not-you {
+		display: block;
+		bottom: 0;
+		margin-bottom: -60px;
```
This looks a bit fragile. Have we tested well on all possible screen sizes?

enejb replied:
Yes, I tested this on and adjusted the value to -40px now. So that we don't end up with a vertical scroll on smaller screens (iPhone 5)

tyxla commented on client/blocks/signup-form/index.jsx:
```diff
@@ -965,6 +968,18 @@ class SignupForm extends Component {
+				<ContinueAsUser
+					redirectPath={ this.props.redirectToAfterLoginUrl }
+					onChangeAccount={ () => {
+						// todo: Add track this event!
```
Are we planning to address this TODO in this PR?

enejb replied:
Done we should be tracking things now.

tyxla commented on client/blocks/signup-form/index.jsx:
```diff
@@ -965,6 +968,18 @@ class SignupForm extends Component {
+			return (
+				<ContinueAsUser
+					redirectPath={ this.props.redirectToAfterLoginUrl }
+					onChangeAccount={ () => {
```
Could we move this function to a separate method of the component?

enejb replied:
Done. Added a new handle function

tyxla commented on client/signup/config/flows.js:
```diff
@@ -197,7 +198,13 @@ const Flows = {
+			// Remove the user step unless the user has just completed the step
+			// and then clicked the back button.
+			if ( ! param && ! detectHistoryNavigation.loadedViaHistory() ) {
+				flow = removeUserStepFromFlow( flow );
```
Can we confirm that this won't break something for an existing signup flow?

enejb replied:
`flow = removeUserStepFromFlow( flow );` was there before. 
It used to be always applied. To test that it still works as expected. 
I create the user and then just visit `/start/user` it should then redirect me to => `/start/domains` 

I am not sure how else to test this?

p-jackson replied:
There is also flows like `p2` and `p2-new` where the user step isn't the first step. I briefly tried them and the step exclusion still seems to work as expected.

tyxla commented on client/signup/steps/user/index.jsx:
```diff
@@ -118,6 +119,20 @@ export class UserStep extends Component {
+			url.search = search_params.toString();
+			// Redirect to itself and append ?user_completed
+			window.location.replace( url.toString() );
+			return;
```
We don't seem to need this `return`.

enejb replied:
Done removed the return.

tyxla commented on client/signup/steps/user/index.jsx:
```diff
@@ -118,6 +119,20 @@ export class UserStep extends Component {
+			// It looks like the user just completed the User Registartion Step
+			// And clicked the back button. Lets redirect them to the this page but this time they will be logged in.
+			const url = new URL( window.location );
+			const search_params = url.searchParams;
```
Why do we need a new variable for `search_params`? Also, let's keep in mind the coding standards and the variable naming conventions (it should be `searchParams` if we decide to keep it).

enejb replied:
I updated the variable to use `searchParams`. 
Not sure if there is a better way to write the same thing though.

tyxla commented on client/signup/steps/user/index.jsx:
```diff
@@ -350,6 +369,12 @@ export class UserStep extends Component {
 
+	userCreationCompletedWithAndHasHistory( props ) {
+		return (
+			props.step && 'completed' === props.step.status && detectHistoryNavigation.loadedViaHistory()
```
We could simplify this by using optional chaining:

```suggestion
			'completed' === props.step?.status && detectHistoryNavigation.loadedViaHistory()
```

tyxla reviewed (commented):
This appears to be testing well! 

I've left some feedback, and not sure if I can add much more from a framework perspective.

Could we perhaps ask someone more familiar with signup and onboarding, like @Automattic/ganon to provide a more thorough review?

enejb committed:
Merge branch 'trunk' into fix/start-user-logged-in

enejb committed:
Simplfy optional chaining

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

enejb committed:
Fix lint error

enejb committed:
removed not needed function return

enejb committed:
Update variable code style

enejb committed:
Add handleOnChangeAccount function + track click

enejb committed:
Continue as user Sign out text variation

enejb committed:
Improve the visibility on iPhone 5

This makes it so that iPhone 5 screen size doesn't have to scroll

enejb committed:
Add comment to explain the rational of the magic numbers.

a8ci18n commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -50,41 +56,60 @@ function ContinueAsUser( { currentUser, redirectUrlFromQuery, onChangeAccount }
 	// the user can notice.
 
+	const notYouText = isSignUpFlow
+		? translate( 'Not you?{{br/}} Sign out or log in with {{link}}another account{{/link}}', {
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **27** times:
`translate( 'Not you?{{br/}}Log in with {{link}}another account{{/link}}' )` *ES Score: 6*

tyxla commented on client/blocks/login/continue-as-user.jsx:
```diff
@@ -50,41 +56,60 @@ function ContinueAsUser( { currentUser, redirectUrlFromQuery, onChangeAccount }
+		? translate( 'Not you?{{br/}} Sign out or log in with {{link}}another account{{/link}}', {
+				components: {
+					br: <br />,
+					link: (
```
Should we declare this component separately? It seems to be duplicated across the 2 strings.

enejb replied:
Great catch!

tyxla commented on client/blocks/signup-form/index.jsx:
```diff
@@ -957,6 +965,16 @@ class SignupForm extends Component {
+				<ContinueAsUser
+					redirectPath={ this.props.redirectToAfterLoginUrl }
+					onChangeAccount={ this.handleOnChangeAccount }
+					isSignUpFlow={ true }
```
The `={ true }` is redundant.

```suggestion
					isSignUpFlow
```

tyxla reviewed (approved):
I don't have much to add regarding code here 👍  This looks like solid work.

I'd defer the final review to someone from @Automattic/ganon as it falls under the signup / onboarding experience.

enejb committed:
remove the redundent attribute.

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

p-jackson commented on client/signup/steps/user/index.jsx:
```diff
@@ -367,6 +385,10 @@ export class UserStep extends Component {
 		return this.props.step && 'completed' === this.props.step.status;
 	}
 
+	userCreationCompletedWithAndHasHistory( props ) {
```
Would this be clearer without the `With`?
Am I interpreting this identifier correctly if I insert a pause like this? "user creation completed, and has history"

enejb replied:
good call

enejb committed:
DRY translation components

p-jackson commented on client/signup/steps/user/index.jsx:
```diff
@@ -135,6 +136,19 @@ export class UserStep extends Component {
 		}
 	}
 
+	componentDidUpdate() {
```
Should this be in `componentDidMount` instead of on each prop update? Feels like we'd only want to potentially redirect on first mount.

enejb replied:
I tried that and it didn't quite work as expected for me. 
It seems that in some cases the component is already mounted and this part gets skipped.

p-jackson reviewed (commented):
I wasn't able to reproduce the issue seen in #57287. Following the login email links I never found myself on a user step with already filled in details.

I've done some testing and it seems to be working as designed. I am finding it quite confusing and tricky to reason about though, especially the use of the `detectHistoryNavigation.loadedViaHistory()` helper. Does this mean there's an intentional difference between the behaviour of using the browser back button and the back button in the framework UI? This seems quite confusing to me.

You may have already investigated this, but can the `.status === 'completed'` flag  be used instead of adding a new `?user_completed` query param? We're not always good at propagating query parameters between steps, so it might be better to rely on the data in the redux store.

enejb committed:
Merge branch 'fix/start-user-logged-in' of github.com:Automattic/wp-calypso into fix/start-user-logged-in

enejb committed:
remove userCreationCompletedWithAndHasHistory since

We don't need this condition any more since componentDidMount did the trick

enejb committed:
Revert "remove userCreationCompletedWithAndHasHistory since"

This reverts commit ead51c7ccc54c33a2eede5cff57f39d9bf6727e1.

enejb committed:
Update the method name to remove With

enejb commented:
Thanks for the review and testing @p-jackson and @tyxla 

I think this PR is ready to go. 

> I've done some testing and it seems to be working as designed. I am finding it quite confusing and tricky to reason about though, especially the use of the detectHistoryNavigation.loadedViaHistory() helper. Does this mean there's an intentional difference between the behaviour of using the browser back button and the back button in the framework UI? This seems quite confusing to me.

1. We removed the back button from the domain screen on mobile since users are pretty used to using the browser back button. Testing the framework UI back button on the desktop eventually leads me to wordpress.com/log-in page. Which doesn't seem to work as intended. Maybe we can address that in a future PR 🤔 

This PR tries to fix this experience for mobile, by sending a user the screen where it is clear that they are already logged in. 

The reason why we need redirect in js is that in the current flow, even if we just created a new user the browser doesn't have the full awareness that the user is logged in yet. 
Which we get once the page is refreshed but once we refresh the page (redirect to itself) we lose the "user just completed registration" state that was in the redux store. (Which doesn't exist any more so having the state in the url via the query. 

If the user is logged in already and they try the `/start` flow it is expected that they will be redirected to the domains page instead of the user registration page. 

Hopefully the above makes sense 😅

p-jackson commented:
> The reason why we need redirect in js is that in the current flow, even if we just created a new user the browser doesn't have the full awareness that the user is logged in yet.

That does make sense now thank you. The signup framework is unique in how it's able to deal with the semi-logged state of the user. It's not surprising that the "Continue As User" screen only works after a full page refresh.

I haven't found any issues with this change in my testing, so no reason to hold off merging due to known bugs; it all seems to be working well. I am worried about the extra complexity that's being added to achieve this experience, but I'm not in a good position to judge the importance of this change to the mobile experience. Feel free to merge if this is going to be better for mobile web users.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7017513

Hi @enejb, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Not you?{{br/}} Sign out or log in with {{link}}another account{{/link}}`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7017513).
